Add GET image/{id}/tiles/thumbnail endpoint.

manthey opened the pull request:
@brianhelba This is done except for tests.  I've pushed it now in case you need it right away.  I will have tests soon.

brianhelba commented on server/tilesource/base.py:
```diff
@@ -55,6 +69,55 @@ def getTile(self, x, y, z):
     def getTileMimeType(self):
         return 'image/jpeg'
 
+    def getThumbnail(self, width=None, height=None, **kwargs):
```
Can we document the available ``kwargs`` here? We don't even need detail, but a comprehensive list is important.

brianhelba commented on server/tilesource/base.py:
```diff
@@ -15,14 +15,28 @@
 
 from ..models.base import TileGeneralException
 
+# Not having PIL disables thumbnail creation, but isn't fatal
```
This is fine for now, but we should start thinking about how to handle graceful degradation in a more organized way.

brianhelba commented on server/rest.py:
```diff
@@ -167,3 +176,42 @@ def deleteTiles(self, item, params):
+               required=False, dataType='int')
+        .param('encoding', 'Thumbnail output encoding', required=False,
+               enum=['PNG', 'JPEG'], default='PNG')
+        .param('jpegQuality', 'Quality used for generating JPEG images',
```
Let's drop ``jpegQuality`` and ``jpegSubsampling`` from the thumbnail HTTP API for now. Leave them available internally in the model / tilesource.

We will definitely want to expose these in the HTTP API for a more generalized "region cutout" endpoint in the future.

brianhelba commented on server/rest.py:
```diff
@@ -97,6 +105,7 @@ def _parseTestParams(params):
         .errorResponse('ID was invalid.')
         .errorResponse('Read access was denied for the item.', 403)
     )
+    @access.cookie
```
Why do we need cookie access here? The metadata endpoint should normally only be called via Ajax, which will be able to set headers.

The image-returning endpoints might be fetched by the DOM (e.g. in ``<img src=``), so they must allow cookie auth.

brianhelba commented:
See my inline comments.

Otherwise, this looks good and is quite a thoughtful implementation. Thanks @manthey .

manthey commented:
@brianhelba I addressed all of your comments and added testing.

I removed cookie access.  If we change our minds (because we want to fetch the thumbnails via `<img>` tags), we can always put it back.

dgutman commented:
Oh we will almost definitely be injecting at least thumbnails using the
<IMG> tag for image galleries.,

On Fri, Mar 18, 2016 at 9:11 AM David Manthey <notifications@github.com>
wrote:

> @brianhelba <https://github.com/brianhelba> I addressed all of your
> comments and added testing.
>
> I removed cookie access. If we change our minds (because we want to fetch
> the thumbnails via <img> tags), we can always put it back.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/DigitalSlideArchive/large_image/pull/34#issuecomment-198348679>
>

brianhelba commented:
@manthey Sorry, I was slightly ambiguous in my last comment about cookies. I was asking why you added the ``@access.cookie`` decorator to the ``getTilesInfo`` endpoint, which just returns JSON metadata. I put an inline comment right next to the change, but my pronouns were unclear.

As @dgutman said, ``getTilesThumbnail`` definitely needs cookie access.

brianhelba commented:
LGTM. Thanks @manthey .
